eventstreamrpc/source/EventStreamClient.cpp

```diff
@@ -0,0 +1,1226 @@
+                }
+                m_mutex.unlock();
+            }
+        }
+
+        template <typename T> void ProtectedPromise<T>::SetValue(const T &lhs) noexcept
+        {
+            if (m_mutex.try_lock())
+            {
+                if (!m_fulfilled)
+                {
+                    m_promise.set_value(lhs);
+                    m_fulfilled = true;
+                }
+                m_mutex.unlock();
+            }
+        }
+
+        template <typename T> std::future<T> ProtectedPromise<T>::GetFuture() noexcept
+        {
+            return m_promise.get_future();
+        }
+
+        class EventStreamCppToNativeCrtBuilder
+        {
+          private:
+            friend class ClientConnection;
+            friend class ClientContinuation;
+            static int s_fillNativeHeadersArray(
+                const Crt::List<EventStreamHeader> &headers,
+                struct aws_array_list *headersArray,
+                Crt::Allocator *m_allocator = Crt::g_allocator)
+            {
+                AWS_ZERO_STRUCT(*headersArray);
+                /* Check if the connection has expired before attempting to send. */
+                int errorCode = aws_event_stream_headers_list_init(headersArray, m_allocator);
+
+                if (!errorCode)
+                {
+                    /* Populate the array with the underlying handle of each EventStreamHeader. */
+                    for (auto &i : headers)
+                    {
+                        errorCode = aws_array_list_push_back(headersArray, i.GetUnderlyingHandle());
+
+                        if (errorCode)
+                        {
+                            break;
+                        }
+                    }
+                }
+
+                return errorCode;
+            }
+        };
+
+        ClientConnection::ClientConnection(ClientConnection &&rhs) noexcept : m_lifecycleHandler(rhs.m_lifecycleHandler)
+        {
+            *this = std::move(rhs);
+        }
+
+        ClientConnection &ClientConnection::operator=(ClientConnection &&rhs) noexcept
+        {
+            m_allocator = std::move(rhs.m_allocator);
+            m_underlyingConnection = rhs.m_underlyingConnection;
+            rhs.m_stateMutex.lock();
+            m_clientState = rhs.m_clientState;
+            rhs.m_stateMutex.unlock();
+            m_lifecycleHandler = rhs.m_lifecycleHandler;
+            m_connectMessageAmender = rhs.m_connectMessageAmender;
+            m_connectAckedPromise = std::move(rhs.m_connectAckedPromise);
+            m_closedPromise = std::move(rhs.m_closedPromise);
+            m_onConnectRequestCallback = rhs.m_onConnectRequestCallback;
+
+            return *this;
+        }
+
+        ClientConnection::ClientConnection(
+            ConnectionLifecycleHandler &connectionLifecycleHandler,
+            Crt::Allocator *allocator) noexcept
+            : m_allocator(allocator), m_lifecycleHandler(connectionLifecycleHandler)
```

Comment:
Probably need to initialize a few more states. At the very least, m_clientState needs to reflect DISCONNECTED and m_underlyingConnection should be nullptr.